### ISSUE TITLE
Design Sprite class and interface

### ISSUE THREAD

**Issue by blacktm** [ruby2d/ruby2d#33]
An [initial implementation](https://github.com/ruby2d/ruby2d/blob/v0.3.0/lib/ruby2d/sprite.rb) has been created for `Sprite`, but the interface hasn't really been "designed," and certainly not yet finalized. In this issue, we can discuss what an ideal sprite DSL might look like and how to implement it. Ideas welcome!

**Comment by KillaPL** [comment]
I will be more than happy to suggest an interface, but I'm gonna need more informations about what sprite is in that context. I come from web development background, and I don't think you mean what I know as sprite; ). Maybe some sources in web?

**Comment by andrewhavens** [comment]
Currently the Sprite API is as follows:
```ruby
sprite = Sprite.new(x, y, "path/to/sprite/sheet")
sprite.start(x, y, width, height) # this sets the frame to be displayed when no animations are playing
sprite.add(animation_name: [
  [x, y, width, height, duration_in_frames], # frame 1
  [x, y, width, height, duration_in_frames], # frame 2
  # etc ...
])
sprite.animate(:animation_name) # start the animation
sprite.reset #  stop the animation and return to the frame specified as the `start` frame
```

I would like to propose that the Text, Image, Sprite start with the path/text of the thing that they are referring to, followed by optional keyword args, in order to achieve some consistency. Some examples of what I mean:

```ruby
# Current:
Sound.new("media/boom.wav")
Text.new(text: "GAME OVER", size: 80, font: "media/mario_font.ttf")
Image.new(path: "media/cloud.png", x: -20, y: -50, z: 100)
Sprite.new(0, 0, "media/sprite_sheet.png")

# Proposed:
Sound.new("media/boom.wav")
Text.new("GAME OVER", size: 80, font: "media/mario_font.ttf")
Image.new("media/cloud.png", x: -20, y: -50, z: 100)
Sprite.new("media/sprite_sheet.png", x: 0, y: 0)
```

Regarding the Sprite animation API, there were two methods that I didn't understand at first: `start` and `reset`. It wasn't clear to me that `start` meant that it would be the frame that is displayed when no other animations are running, and it wasn't clear to me that `reset` would stop the current animation. Maybe we could change it to something like this:

```ruby
# Before:
sprite = Sprite.new(0, 0, "media/sprite_sheet.png")
sprite.start(0, 0, 150, 215)
sprite.add(walking: [
  # x, y, width, height, duration in frames
  [150, 0, 150, 215, 8],
  [300, 0, 150, 215, 8],
])
sprite.animate(:walking)
sprite.reset

# After:
sprite = Sprite.new("media/sprite_sheet.png")
sprite.default(0, 0, 150, 215)
sprite.add(:walking, [
  { x: 150, y: 0, w: 150, h: 215, d: 8 }
  { x: 300, y: 0, w: 150, h: 215, d: 8 }
])
sprite.start(:walking)
sprite.stop
```

I'm not so sure about using a hash for defining the frames (but it's also harder to remember the order of positional arguments without a comment to self), or defining the "idle frame" using `default`, but I know that I like saying "start walking" instead of "animate walking" and "stop" instead of "reset". Also, I think using a symbol as the first argument to `add` could make it more clear that it's the name of the animation.

What do you think?

**Comment by blacktm** [comment]
Hey Andrew! Excited to finally get this done. I _really_ like your idea of having Text, Image, and Sprite (and future classes that are primarily representations of assets) starting with a fixed param, followed by optional kwargs. I don't know why I was thinking in positional or kwargs only — this is a great example of how both can be used well together.

Regarding animation, I was writing up an API spec during the holidays around the same time you posted this and it was surprisingly similar, so that's encouraging! I was thinking about the various use cases we would run into, using [Kenney's awesome assets](http://kenney.nl) to sketch out some examples. Note that many of these ideas take after the "convention over configuration" approach that is becoming a common philosophy in Ruby, at least by default (we can always add _a lot_ more optional configs). Let me know if you can think of other example use cases as well.

**1. Sprite with fixed clipping rectangle size in horizontal order and looping**

Here's an example with a rotating coin, which will loop through each frame from the right to left at multiples of 86 pixels, advancing after 300 milliseconds.

![coin](https://user-images.githubusercontent.com/887429/34644218-f2809318-f2e6-11e7-925e-13c468776653.png)

```ruby
coin = Sprite.new(
  "coin.png",
  clip_width: 86,
  time: 300,
  loop: true
)

coin.play
```

**2. Fixed clipping rectangle (same as 1) with single play through**

Here's an example with an explosion.

![boom](https://user-images.githubusercontent.com/887429/34644221-fbea55a6-f2e6-11e7-812c-377ba855617c.png)

```ruby
boom = Sprite.new(
  "boom.png",
  clip_width: 120,
  time: 300
)

boom.play
```

**3. Fixed clipping rectangle with multiple animations**

This kind of thing would be used for a character sprite. Default would be set at frame 0. (Also, should we use array-style indexing, or start at `1`?) In the spirit of "make the simple things easy," each animation here uses the global config (like time/duration) and animations are just an array of frame indices, but a hash could be used to describe each frame specifically (with differing attributes) like in the the "sprite atlas" example that follows.

![hero](https://user-images.githubusercontent.com/887429/34644224-041f713e-f2e7-11e7-8a09-ea5ed158b1ed.png)

```ruby
hero = Sprite.new(
  "hero.png",
  clip_width: 72,
  time: 300,
  animations: {
    walk_right: [1, 2],
    climb: [3, 4],
    cheer: [5, 6]
  }
)

# Play "walk right" animation on loop
hero.play :walk_right, :loop

# Pause the animation at current frame
hero.pause

# Stop the animation and reset to default frame
hero.stop

# Alternatively, this could go back to default frame
hero.reset

# Stop animation on frame 0 (two ideas)
hero.stop :frame 5
hero.stop_at 5
```

**4. [Sprite atlas](https://en.wikipedia.org/wiki/Texture_atlas) with different sized clipping rectangles or different frame durations**

This is more of an explicit way of defining the details of the animation frames. This would be necessary if using a sprite atlas, where the frames are all over the place, different sizes, etc. This config could also be stored in a file (as JSON or something) read in with the sprite image ([texture packing formats](http://doc.starling-framework.org/v2.2/starling/textures/TextureAtlas.html) are pretty common).

```ruby
thing = Sprite.new(
  "thing.png",
  animations: {
    some_action: [
      {
        x: 0, y: 0,
        width: 25, height: 50,
        time: 300
      },
      {
        x: 25, y: 50,
        width: 50, height: 75,
        time: 200
      },
      {
        x: 50, y: 100,
        width: 25, height: 25,
        time: 100
      }
    ]
  }
)

thing.play :some_action
```

The frame details could be optional and would choose global values if specified. In this example, the clipping rectangle size is the same, but each frame of the animation is a different duration (gets faster).

```ruby
thing = Sprite.new(
  "thing.png",
  clip_width: 25
  animations: {
    some_action: [
      { time: 300 },
      { time: 200 },
      { time: 100 }
    ]
  }
)

thing.play :some_action
```

**Comment by alexdovzhanyn** [comment]
This seems solid. Is there a way to interface with the sprite image sizing? An issue I had when working with sprite was that my files were 30×30, but my screen resolution was was high, so I needed to upscale the sprites by a factor of 5, for example. I wasnt able to do so, because 1) there wasnt a way to set the height/width of the sprite after creating it and 2) there was no way of retrieving those attributes from a sprite instance. If I'm overlooking something let me know, but I think that would be an importand addition to the sprite api

**Comment by blacktm** [comment]
@alexdovzhanyn Yes, resizing should be possible — natively, we're storing both the [sprite width and height](https://github.com/simple2d/simple2d/blob/20f2aa0a06b742ddd309aa7bb1f40c52bc291a33/include/simple2d.h#L268) (in addition to the clipping rectangle), and furthermore it's [composed of an image](https://github.com/simple2d/simple2d/blob/20f2aa0a06b742ddd309aa7bb1f40c52bc291a33/include/simple2d.h#L254), which has both a width/height (for resizing) and the original size. Now, I don't think it's in the Ruby API yet, but we can spec it out here, see what the gaps are.

**Comment by andrewhavens** [comment]
@blacktm Your API looks good to me! I like the new simple way of defining animations by frame indices. I also think the expanded "sprite atlas" version is good for those times where you need more control. While working with cocos2d, I used the [TexturePacker](https://www.codeandweb.com/texturepacker) app to generate a sprite sheet, so I think supporting a common format like that would be important. There's also another feature closely related to sprites called Tile Maps. I used an app called [Tiled](http://www.mapeditor.org/) to generate my Tile Map. We should support that too, but I think that's a separate issue.

The only thing that I thought was weird (or didn't understand) was this:
```
# Stop animation on frame 0 (two ideas)
hero.stop :frame 5
hero.stop_at 5
```
I'm not sure that we need this, and I don't understand which frame I would be stopped at. Frame 5 would seem to imply that it was the 5th frame of the current animation, but your comment mentions frame 0. I guess I also can't think of a use-case for this that couldn't be handled in some other way.

Regarding resizing/scaling...I've seen other libraries use `scale` as their way of resizing. We could allow a scale argument like this:
```
image = Image.new("image.png", scale: 2) # twice the original size
sprite = Sprite.new("sprite.png", scale: 0.5) # half the original size
```

**Comment by shawn42** [comment]
A lot of this looks great!

It reminds me of similar stuff I did (back in the day) on gamebox:
[code](https://github.com/shawn42/killbox/blob/master/src/actors/player.rb#L8-L19)

One thing I like about Gosu/Gamebox is the ability to specify rows and columns instead of exact pixel width and height.

I really like the Sprite atlas stuff too! It allows the flexibility if you need it, but can have a simple interface if you don't.

**Comment by masondesu** [comment]
@blacktm before commenting, I put together a demo with a number of animations using the current `Sprite` class, and agree with the other feedback here that the new suggested API is a big improvement. Especially glad that a single well-designed class will handle all of the use cases you mentioned.

**Comment by blacktm** [comment]
Thanks for your feedback everyone! I'll get to it and open a working PR for this.

@andrewhavens Yea, the "stop on frame" was a typo, should've been frame 5. Not sure how useful that would be anyways. Definitely like the scale idea, since you'd probably be keeping the width/height proportions fixed anyways.

@shawn42 Thanks for sharing that example — nice use of ranges there 🤔

@masondesu Awesome, that's great to hear!

**Comment by blacktm** [comment]
Alright, I think this (#99) is ready for a code review if anyone is interested.

**Comment by blacktm** [comment]
Merged in #99, thanks y'all!
